cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -2078,6 +2078,293 @@ class ArrayRanker : public TypeVisitor {
+        if (has_null_like_values<typename ArrayType::TypeClass>::value) {
+          PartitionNullsOnly<StablePartitioner>(nulls_begin, nulls_end,
+                                                ChunkedArrayResolver(arrays), null_count,
+                                                null_placement_);
+        }
+      };
+      auto merge_non_nulls = [&](uint64_t* range_begin, uint64_t* range_middle,
+                                 uint64_t* range_end, uint64_t* temp_indices) {
+        MergeNonNulls<ArrayType>(range_begin, range_middle, range_end, arrays,
+                                 temp_indices);
+      };
+
+      MergeImpl merge_impl{null_placement_, std::move(merge_nulls),
+                           std::move(merge_non_nulls)};
+      // std::merge is only called on non-null values, so size temp indices accordingly
+      RETURN_NOT_OK(merge_impl.Init(ctx_, indices_end_ - indices_begin_ - null_count));
+
+      while (sorted.size() > 1) {
+        auto out_it = sorted.begin();
+        auto it = sorted.begin();
+        while (it < sorted.end() - 1) {
+          const auto& left = *it++;
+          const auto& right = *it++;
+          DCHECK_EQ(left.overall_end(), right.overall_begin());
+          const auto merged = merge_impl.Merge(left, right, null_count);
+          *out_it++ = merged;
+        }
+        if (it < sorted.end()) {
+          *out_it++ = *it++;
+        }
+        sorted.erase(out_it, sorted.end());
+      }
+    }
+
+    DCHECK_EQ(sorted.size(), 1);
+    DCHECK_EQ(sorted[0].overall_begin(), indices_begin_);
+    DCHECK_EQ(sorted[0].overall_end(), indices_end_);
+    // Note that "nulls" can also include NaNs, hence the >= check
+    DCHECK_GE(sorted[0].null_count(), null_count);
+
+    auto length = indices_end_ - indices_begin_;
+    ARROW_ASSIGN_OR_RAISE(auto rankings,
+                          MakeMutableUInt64Array(uint64(), length, ctx_->memory_pool()));
+    auto out_begin = rankings->GetMutableValues<uint64_t>(1);
+    uint64_t rank;
+
+    switch (tiebreaker_) {
+      case RankOptions::Dense: {
+        T curr_value, prev_value{};
+        rank = 0;
+
+        if (null_placement_ == NullPlacement::AtStart && sorted[0].null_count() > 0) {
+          rank++;
+          for (auto it = sorted[0].nulls_begin; it < sorted[0].nulls_end; it++) {
+            out_begin[*it] = rank;
+          }
+        }
+
+        for (auto it = sorted[0].non_nulls_begin; it < sorted[0].non_nulls_end; it++) {
+          // Below code wasn't working for string specialization as -> value returned a buffer
+          // but T is a basic_string_view
+          // using ScalarType = typename TypeTraits<InType>::ScalarType;
+          // auto scalar = std::dynamic_pointer_cast<ScalarType>(
+          //   chunked_array_.GetScalar(*it).ValueOrDie());
+          // curr_value = scalar->value;
+          // TODO: can we use chunk_resolver_ from chunked array externally?          
+          if (*it >= 2) {
```

Comment:
This is hacked together just to work for the provided test case but obviously needs a better implementation. I wasn't sure what the best approach was to get the logical value from a particular index in a ChunkedArray. Alternately we can bypass the ChunkedArray and try to use the `arrays` variable, but I don't know that that would be any cleaner

Reply:
My current best guess is to execute `take` with the indices and iterate over that result. I don't know that I've seen that pattern internally yet so happy to go a different route if that is too heavy handed